### ISSUE TITLE
What's the performance profile of Microsoft365DSC? Can it be used to manage hundreds of tenants?

### ISSUE THREAD

**Issue by markoueis** [microsoft/Microsoft365DSC#3270]
This is more of a higher architecture and performance question. 

If this is monitoring every 15 minutes (I assume that's changeable), what kind of resource utilization can we expect? Can a single deployment of this be used to manage hundreds of Office365/teams/azuread/etc tenants? Can it be used to manage multiple enterprises?

Thank you

**Comment by andikrueger** [comment]
Performance would  highly depend on the number of resources (settings) you would like to monitor. Basically every resource you would like to monitor would be at best case one request to get the data from the tenant. 

The 15 minutes are the default setting of the local configuration manager that would run your configuration and the consistency tests. 

From my experience, large configurations can take up to hours to be applied and monitored within a tenant. 

In your scenario, I would go for an architectural design which would be scalable: One runner per tenant or an environment that would allow parallelism of various powershell processes.

**Comment by markoueis** [comment]
Thank you Andi for that quick response. What did you have in mind when you say an environment that would allow parallelism of various PowerShell processes? Windows can run PowerShell processes in parallel. Could we run these in like azure functions for example? Does PowerShell DSC supporting something like that?

As for the VM/container approach, technically, then, it is possible for one VM to manage multiple tenants. Would it be fair to say we would have to test to how many tenants one reasonably powered VM (4 vcpu, 16g ram) can handle before we start to run into resource utilization problems, but we should be able to do dozens ok? I just don't want to pursue this if we will have problems with dozens to begin with.

**Comment by andikrueger** [comment]
There might be one decision to make first: are you planning to run the validation by using the local configuration manager (LCM) or are you creating a system that would be capable of running stuff similar to LCM. 

if you would go with LCM, you would create a 1:1 relationship between a configuration and a windows system. This 1:1 relationship would be reasonable in regard to m365 as well - in your special scenario it would cause too much resource consumption. This would lead to configurations that would not only target on tenant.

Every resource asks for authentication parameters. If you’d like to reduce the number of windows systems, you could combine multiple tenants in one configuration and run it within LCM. 

LCM on the other hand, does not really enable you to run processes in parallel. 

Running the process in azure functions would be doable. You would need to think of a solution on how to pass the required properties to the test-targetresource function of every resource’s module. Same would allow parallelism on a windows system with multiple powershell processes. 

Either approach might be bound to another challenge: reporting and acting on a drift.

**Comment by William-Francillette** [comment]
For large deployment such a 100 of tenants, I would use windows containers and kubernetes.

You could split each customer/tenant into worker nodes (1:1) and within your worker nodes split your M365DSC workload into multiple pods such a pod for AAD, a pod for SP and so forth 

For a single or a few customer/tenant then docker should be ok

I was labing with M365DSC and docker and outcomes were very promising. 
Splitting the workload into multiple containers greatly improves the performance (parallel run) and the maintenance 
My environment was nested virtualization (hyper-V) and certificate authentication.

I am planning to write a blog post about the topic, I have just been caught in resource development and exam training 😎

**Comment by andikrueger** [comment]
Just a quick note on windows containers on docker - this is what I did some testing with half a year ago. It was not as lightweight as one would wish. The image was way to big and execution were slow. 
M365DSC did not play too well in PS 7.x back then. It would be my recommendation to stick with PS 5.1 as for now. 

@William-Francillette looking forward to your insights on this topic. 

Splitting up the various resources and customers is highly recommended. This would allow for a greater degree of parallelism.

**Comment by William-Francillette** [comment]
I agree with you my windows-core image was above 6gb and performance were slow but i was using my laptop and a small windows 10 vm (1 core 4gb ram). I was also testing on PS 5.1

Hopefully DSC support for Linux will be improved and hope we can run M365DSC on PS 7.1 and a Linux box
This would be so great for us

**Comment by andikrueger** [comment]
It is also slow for Azure Containers. This is where I ran my tests. It took up to 15 mins to get the container ready. There are other services like runbooks or functions which would allow a more instant run of PowerShell scripts.
